src/test/assembly/asm/powerpc-types.rs

```diff
@@ -0,0 +1,105 @@
+// min-llvm-version: 10.0.1
+// assembly-output: emit-asm
+// compile-flags: --target powerpc-unknown-linux-gnu
+// needs-llvm-components: powerpc
+
+#![feature(no_core, lang_items, rustc_attrs, repr_simd)]
+#![crate_type = "rlib"]
+#![no_core]
+#![allow(asm_sub_register, non_camel_case_types)]
+
+#[rustc_builtin_macro]
+macro_rules! asm {
+    () => {};
+}
+#[rustc_builtin_macro]
+macro_rules! concat {
+    () => {};
+}
+#[rustc_builtin_macro]
+macro_rules! stringify {
+    () => {};
+}
+
+#[lang = "sized"]
+trait Sized {}
+#[lang = "copy"]
+trait Copy {}
+
+type ptr = *const i32;
+
+impl Copy for i8 {}
+impl Copy for u8 {}
+impl Copy for i16 {}
+impl Copy for i32 {}
+impl Copy for i64 {}
+impl Copy for f32 {}
+impl Copy for f64 {}
+impl Copy for ptr {}
+extern "C" {
+    fn extern_func();
+    static extern_static: u8;
+}
+
+// Hack to avoid function merging
+extern "Rust" {
+    fn dont_merge(s: &str);
+}
+
+macro_rules! check { ($func:ident, $ty:ty, $class:ident, $mov:literal) => {
+    #[no_mangle]
+    pub unsafe fn $func(x: $ty) -> $ty {
+        dont_merge(stringify!($func));
+
+        let y;
+        asm!(concat!($mov," {}, {}"), out($class) y, in($class) x);
+        y
+    }
+};}
+
+macro_rules! check_reg { ($func:ident, $ty:ty, $rego:tt, $regc:tt, $mov:literal) => {
+    #[no_mangle]
+    pub unsafe fn $func(x: $ty) -> $ty {
+        dont_merge(stringify!($func));
+
+        let y;
+        asm!(concat!($mov, " ", $rego, ", ", $rego), lateout($regc) y, in($regc) x);
+        y
+    }
+};}
+
+// CHECK-LABEL: reg_i32:
+// CHECK: #APP
+// CHECK: mr {{[0-31]+}}, {{[0-31]+}}
+// CHECK: #NO_APP
+check!(reg_i32, i32, reg, "mr");
+
+// CHECK-LABEL: reg_i32_nz:
+// CHECK: #APP
+// CHECK: mr {{[1-31]+}}, {{[1-31]+}}
+// CHECK: #NO_APP
+check!(reg_i32_nz, i32, reg_nonzero, "mr");
+
+// CHECK-LABEL: reg_f64:
+// CHECK: #APP
+// CHECK: fmr {{[0-31]+}}, {{[0-31]+}}
+// CHECK: #NO_APP
+check!(reg_f64, f64, freg, "fmr");
+
+// CHECK-LABEL: reg_r0:
+// CHECK: #APP
+// CHECK: mr 0, 0
+// CHECK: #NO_APP
+check_reg!(reg_r0, i32, "0", "0", "mr");
+
+// CHECK-LABEL: reg_r18:
+// CHECK: #APP
+// CHECK: mr 18, 18
+// CHECK: #NO_APP
+check_reg!(reg_r18, i32, "18", "18", "mr");
+
+// CHECK-LABEL: reg_f0:
+// CHECK: #APP
+// CHECK: fmr 0, 0
+// CHECK: #NO_APP
+check_reg!(reg_f0, f32, "0", "f0", "fmr");
```

Comment:
Done - except I ran into a weird error with LLVM.

This code:
```
// CHECK-LABEL: reg_f64_f0:
// CHECK: #APP
// CHECK: fmr 0, 0
// CHECK: #NO_APP
check_reg!(reg_f64_f0, f64, "0", "f0", "fmr");
```

Generates this error:
```
LLVM ERROR: Do not know how to expand this operator's operand!
```

Which is weird, because I have the following test that works okay:
```
// CHECK-LABEL: reg_f64:
// CHECK: #APP
// CHECK: fmr {{[0-31]+}}, {{[0-31]+}}
// CHECK: #NO_APP
check!(reg_f64, f64, freg, "fmr");
```

Reply:
Can you create a minimal test file to reproduce this and then dump the LLVM IR with --emit llvm-ir?

Reply:
Sure - can you give a quick command to dump the IR? I'm not sure how to do it with a development compiler.

Reply:
Seems to be a bug in LLVM that's fixed in LLVM master: https://godbolt.org/z/Mnn9E6G6c

Reply:
Ah, dang. What's the recommended path forward then?

Reply:
I believe [this](https://reviews.llvm.org/rG4e127bce2d1133ba95a551d69bd0e8fc3b4f9e71) commit needs to be cherry-picked into our LLVM branch in https://github.com/rust-lang/llvm-project/. Would you like to send a PR for it?

Reply:
I suppose it doesn't matter who submits a PR - if you get to it first, go for it :)